### PR TITLE
STY: Update macosx zoom rect styling

### LINKED ISSUE #23777: [ENH] Interactive Zoom Rectangle Color Review for MACOSX backend
Interactive zoom rectangle is not visible on dark backgrounds for MACOSX backend.

Seems like this has been fixed when using QT backend (issue [#16798](https://github.com/matplotlib/matplotlib/pull/16798)), but can't find any solution / workaround for MACOSX backend.

Light Background

Dark Background

### PR DESCRIPTION
**Pull request by greglucas**
## PR Summary

Match the zoom box styling of the other backends with alternating white/black paths.

closes #23777

## PR Checklist

**Tests and Styling**
- [N/A] Has pytest style unit tests (and `pytest` passes).
- [N/A] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).

**Documentation**
- [N/A] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).
- [N/A] New plotting related features are documented with examples.

**Release Notes**
- [N/A] New features are marked with a `.. versionadded::` directive in the docstring and documented in `doc/users/next_whats_new/`
- [N/A] API changes are marked with a `.. versionchanged::` directive in the docstring and documented in `doc/api/next_api_changes/`
- [N/A] Release notes conform with instructions in  `next_whats_new/README.rst` or `next_api_changes/README.rst`

### CODE REVIEW

File: src/_macosx.m [hunk 1]
```diff
@@ -1526,16 +1537,17 @@ - (void)otherMouseDragged:(NSEvent *)event { [self mouseDragged: event]; }
 
 - (void)setRubberband:(NSRect)rect
 {
-    if (!NSIsEmptyRect(rubberband)) { [self setNeedsDisplayInRect: rubberband]; }
     rubberband = rect;
-    [self setNeedsDisplayInRect: rubberband];
+    // We can't just set DisplayInRect because the outer paths need to be removed
+    // when shrinking the zoom-box inwards
+    [self setNeedsDisplay: YES];
```
Comment (QuLogic) [inline src/_macosx.m:1543]: You could take the union of the existing `rubberband` with the new `rect` (assuming ObjC has that API)?

Comment (greglucas) [reply src/_macosx.m]: Yep, good call, updated with that now.

File: src/_macosx.m [hunk 2]
```diff
@@ -1330,7 +1330,18 @@ -(void)drawRect:(NSRect)rect
         goto exit;
     }
     if (!NSIsEmptyRect(rubberband)) {
-        NSFrameRect(rubberband);
+        // We use bezier paths so we can stroke the outside with a dash
+        // pattern alternating white/black with two separate paths offset
+        // in phase.
+        NSBezierPath *white_path = [NSBezierPath bezierPathWithRect: rubberband];
+        NSBezierPath *black_path = [NSBezierPath bezierPathWithRect: rubberband];
+        double dash_pattern[2] = {3, 3};
```
Comment (QuLogic) [inline src/_macosx.m:1338]: Should be `CGFloat` technically.
```suggestion
        CGFloat dash_pattern[2] = {3, 3};
```

File: src/_macosx.m [hunk 3]
```diff
@@ -1526,9 +1537,11 @@ - (void)otherMouseDragged:(NSEvent *)event { [self mouseDragged: event]; }
 
 - (void)setRubberband:(NSRect)rect
 {
-    if (!NSIsEmptyRect(rubberband)) { [self setNeedsDisplayInRect: rubberband]; }
+    // The space we want to redraw is a union of the previous rubberband
+    // with the new rubberband and then expanded (negative inset) by one
```
Comment (QuLogic) [inline src/_macosx.m:1541]: Just checking if this is in logical space or physical space (i.e., should be `one` or `one * display ratio`)?

**Comment by oscargus** [comment]
Two things (I cannot test or judge the code...):

1. The (some) other backends draw one solid line and one dashed on top. It may be worth checking if that improves or degrades performance. (It is primarily done because it was hard to get an offset in those, though.)
2. This can possibly go in 3.6.3 as it is in some sense a bug fix. The other backends went into 3.6.2.

File: src/_macosx.m (hunk 3 above)
Comment (greglucas) [reply src/_macosx.m:1541]: On HiDPI, this still clears the region and I'm not sure we want to bring in the deviceScale/contextRegion if we don't absolutely need to here. The linewidth is drawn in points and centered on the rectangle area, so the default linewidth of 1 expands 0.5 points on either side of the rect. (linewidth section from here
https://developer.apple.com/library/archive/documentation/Cocoa/Conceptual/CocoaDrawingGuide/Paths/Paths.html#//apple_ref/doc/uid/TP40003290-CH206-BBCDEJGD)

**Comment by greglucas** [comment]
> The (some) other backends draw one solid line and one dashed on top. It may be worth checking if that improves or degrades performance. (It is primarily done because it was hard to get an offset in those, though.)

This is already so fast I don't think it is worth worrying about performance. I also don't want to overlap if we can avoid it so we don't have any additional weird aliasing / blending going on for the overlapping regions.

> This can possibly go in 3.6.3 as it is in some sense a bug fix. The other backends went into 3.6.2.

I see this as an enhancement/improvement and not a bugfix. I really don't think these should be backported, but rather if we want it in sooner, a new minor release should happen sooner. The cursor styling and other aesthetic changes like that should be in minor versions not patch IMO.

**Comment by oscargus** [comment]
OK! Makes sense. I was under the impression that something had changed in recent versions, but maybe it is just a consequence of more and more people using dark themes that we get those bug reports now...

**Comment by greglucas** [comment]
ping for a macosx user to review this. It'd be nice to get in for the 3.7 release to keep the backends consistent in their styling.

**Review by jklymak — APPROVED** [review]
This works for me.  The dashes are a bit small on my hiDPI screen compared to a zoom rectangle in Preview.app, but not crazy different.  I'm a little surprised there is not an API for a rubberband already, but this seems to work.

**Comment by tacaswell** [comment]
Confirmed that this looks good axes with white, black, and "0.5" patch colors.

**Comment by greglucas** [comment]
>  The dashes are a bit small on my hiDPI screen compared to a zoom rectangle in Preview.app, but not crazy different

I think the lines should be the same thickness as before, just dashed now. It is easy enough to increase the thickness if desired too. On preview for me, the dash pattern actually rotates around.

**Comment by jklymak** [comment]
> On preview for me, the dash pattern actually rotates around.

Yes, but I wasn't going to ask you to do that.  However, that is what made me think there must be a selector API

**Comment by QuLogic** [comment]
> On preview for me, the dash pattern actually rotates around.

What do you mean by rotates around?

**Comment by jklymak** [comment]
The dash pattern precesses around the selection in a clockwise manner

**Pull request merged**